src/app/services/auth.service.ts

```diff
@@ -66,33 +67,74 @@ and check if you have setup proper "Valid Redirect URIs" and "Web Origins" value
 
     login(): Promise<void> {
         if (this.isEnabled()) {
-            return this.auth.login();
+            return this.keycloak.login();
         } else {
             return Promise.reject('not enabled');
         }
     }
 
+    /**
+   * Return the users realm level roles
+   */
+    public getRealmRoles(): string[] {
+        if (this.keycloak.realmAccess && this.keycloak.realmAccess.roles) {
+            return this.keycloak.realmAccess.roles;
+        }
+        return [];
+    }
+
     logout() {
         if (this.isEnabled()) {
             this.authState.logout();
-            return this.auth.logout();
+            return this.keycloak.logout();
         } else {
             return Promise.reject('not enabled');
         }
     }
 
     getAuthContextProvider(): Promise<AuthContextProvider | undefined> {
         if (this.isEnabled()) {
-            return this.initialized.then( (success) => {
+            return this.initialized.then((success) => {
                 if (success) {
-                    return this.auth.getAuthContextProvider();
+                    return () => {
+                        const tokenUpdate = this.keycloak.updateToken(30) as any;
+                        // Keycloak doesn't use a proper promise. Instead it uses success/error.
+                        return new Promise<AuthContext>((resolve, reject) => {
+                            tokenUpdate.success(() => {
+                                if (this.keycloak.token) {
```

Comment:
This if statement is 3 levels deep. Perhaps lines 102-118 could be extracted into another method?

```ts
const tokenUpdate = this.keycloak.updateToken(30) as any;
return this.getAuthHeaders(tokenUpdate);
```

Reply:
Yeah... That ugly code was moved from js-sdk. See how easy it is now